### ISSUE TITLE
Pyzo 4.18 debugging not working: 'Debugger' object has no attribute 'quitting'

### ISSUE THREAD

**Issue by raymanP** [pyzo/pyzo#1101]
Hi there,

after opening Pyzo and having breakpoints active before executing the first script or a command the debugger raises an exception:
`AttributeError: 'Debugger' object has no attribute 'quitting'`

If having no breakpoints active and execution a command or a script, afterwards breakpoints can be set and working for the following commands / script executions (this is working as a workaround :))

Looking forward for a fix, thank you :)
RaymanP

**Comment by bdieterm** [comment]
Hi @raymanP!

I tried to reproduce your problem but still to no avail.

The `Debugger` class is derived from [`bdb.Bdb`](https://docs.python.org/3/library/bdb.html) from the Python standard library, and that Python code is only executed inside the Python interpreter that is specified in the shell configuration.

Can you post your shell configuration, for example screenshots, or the output of `print(pyzo.config.shellConfigs2)` executed in Pyzo's Logger tool?

Can you also post the initial output of the shell when it is started, for example:
```
Python 3.13.0 (main, Oct 16 2024, 03:10:20) on linux (64 bits).
This is the Pyzo interpreter with integrated event loop for ASYNCIO.
Type 'help' for help, type '?' for a list of *magic* commands.
```

Does the problem reappear after restarting the shell (e.g. via Ctrl+K)?

Is there more information in the full error message of the AttributeError, for example a stack trace?

Does the problem also appear with other Python interpreters configured in the shell configuration?

**Comment by raymanP** [comment]
Hi @bdieterm ,

it occurs with any of the Python interpresters and on all OS systems (Win, Linux, macOS). 
With and without Venv, also after interpreter restart. 

But I found out it occurs with IPython active. 
Here is the terminal in/output:
```
Python 3.12.7 (main, Oct  1 2024, 02:05:46) on darwin (64 bits).
This is the Pyzo interpreter with integrated event loop for ASYNCIO.

Using IPython 8.26.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: 4
Error in callback <bound method PyzoInterpreter.ipython_pre_run_cell_hook of <pyzokernel.interpreter.PyzoInterpreter object at 0x107415310>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 1095bda00, raw_cell="4" store_history=True silent=False shell_futures=True cell_id=None>,),kwargs {}:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File /Applications/pyzo.app/Contents/Resources/source/pyzo/pyzokernel/interpreter.py:1066, in PyzoInterpreter.ipython_pre_run_cell_hook(self, ipython)
   1064 """Hook that IPython calls right before executing code."""
   1065 self.apply_breakpoints()
-> 1066 self.debugger.set_on()

File /Applications/pyzo.app/Contents/Resources/source/pyzo/pyzokernel/debug.py:141, in Debugger.set_on(self)
    138     self._interacting = False
    139     sys.settrace(None)
--> 141 def set_on(self):
    142     """To turn debugging on right before executing code."""
    143     # Reset and set bottom frame

File /Applications/pyzo.app/Contents/Resources/source/pyzo/pyzokernel/debug.py:62, in Debugger.trace_dispatch(self, frame, event, arg)
     59         for i in self.breaks.get(clean_filename, []):
     60             self.set_break(ori_filename, i - offset)
---> 62 return bdb.Bdb.trace_dispatch(self, frame, event, arg)

File /usr/local/Cellar/python@3.12/3.12.7_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/bdb.py:87, in Bdb.trace_dispatch(self, frame, event, arg)
     63 def trace_dispatch(self, frame, event, arg):
     64     """Dispatch a trace function for debugged frames based on the event.
     65 
     66     This function is installed as the trace function for debugged
   (...)
     85     The arg parameter depends on the previous event.
     86     """
---> 87     if self.quitting:
     88         return # None
     89     if event == 'line':

AttributeError: 'Debugger' object has no attribute 'quitting'
Out[1]: 4
```

Here's the config output:
```
>>> print(pyzo.config.shellConfigs2)
[Dict(name='Shell config 5', exe='/usr/local/Cellar/python@3.8/3.8.20/Frameworks
/Python.framework/Versions/3.8/bin/python3.8', gui='auto', pythonPath='', 
projectPath='', scriptFile='', startDir='', startupScript='', argv='', environ='', ipython='yes')]
```

**Comment by bdieterm** [comment]
Hi @raymanP!

Thanks for the detailed and quick answer. I was able to reproduce the problem, for example using Python 3.13.0 with IPython 8.29.0 on Linux.

And I already merged a fix for that problem: see #1102
If you want, you can patch Pyzo yourself by modifying file "pyzo/pyzokernel/interpreter.py" so that you do not have to wait for the next release. Feel free to ask if there are any questions.

Please let us know if this worked for you or if you encounter any problems.

**Comment by raymanP** [comment]
I tried it. Is now working!
Thank you @bdieterm ! :))
